cellpack/autopack/Analysis.py

```diff
@@ -927,6 +1010,335 @@ def PairCorrelationFunction_2D(self, x, y, S, rMax, dr):
+                        all_spilr_scaled[pc, sc],
+                        cbar=False,
+                        xticklabels=False,
+                        yticklabels=False,
+                    )
+                    g_scaled.set_xlabel("Angular coordinates")
+                    g_scaled.set_ylabel(
+                        f"Scaled distance from Nuclear Surface, {pc}_{sc}"
+                    )
+                    g_scaled.invert_yaxis()
+
+                    fig_raw = g_raw.get_figure()
+                    fig_scaled = g_scaled.get_figure()
+                    fig_raw.savefig(
+                        f"{save_dir}heatmap_raw_{pc}_{sc}_{self.ingr_key}",
+                        dpi=300,
+                    )
+                    fig_scaled.savefig(
+                        f"{save_dir}heatmap_scaled_{pc}_{sc}_{self.ingr_key}",
+                        dpi=300,
+                    )
+                    saved_plots += 1
+
+        if get_correlations:
+            print("calculating correlations...")
+            corr_df["packing_id"] = numpy.nan
+            for pc1 in range(num_packings):
+                for sc1 in tqdm(range(num_seeds[pc1])):
+                    for pc2 in range(num_packings):
+                        for sc2 in range(num_seeds[pc2]):
+                            corr_df.loc[f"{pc1}_{sc1}", "packing_id"] = pc1
+                            # do not calculate if:
+                            # a) already calculated
+                            # b) calculating for same packing
+                            if (
+                                not numpy.isnan(
+                                    corr_df.loc[f"{pc1}_{sc1}", f"{pc2}_{sc2}"]
+                                )
+                            ) or ((pc1 == pc2) and (sc1 == sc2)):
+                                continue
+                            corr_df.loc[
+                                f"{pc1}_{sc1}", f"{pc2}_{sc2}"
+                            ] = self.get_spilr_correlations(
+                                all_spilr_scaled[pc1, sc1],
+                                all_spilr_scaled[pc2, sc2],
+                            )
+                            corr_df.loc[f"{pc2}_{sc2}", f"{pc1}_{sc1}"] = corr_df.loc[
+                                f"{pc1}_{sc1}", f"{pc2}_{sc2}"
+                            ]
+            df_packing = corr_df.pop("packing_id")
+            lut = dict(zip(df_packing.unique(), sns.color_palette()))
+            row_colors = df_packing.map(lut)
+            corr_df.fillna(0, inplace=True)
+            g = sns.clustermap(
+                corr_df,
+                row_colors=row_colors,
+                cbar_kws={"label": "spilr correlation"},
+            )
+            g.savefig(f"{self.output_path}spilr_correlation_{self.ingr_key}", dpi=300)
+
+        average_spilr_raw = numpy.nanmean(all_spilr_raw, axis=1)
+        average_spilr_scaled = numpy.nanmean(all_spilr_scaled, axis=1)
+        if save_plots:
```

Comment:
this is an incredibly long function. I think at least this chunk and everything under `if get_correlations` could each be functionalized to `create_and_save_plots` and `get_and_save_correlations`

Reply:
Added helper functions to calculate correlations and to save heatmaps. This also gets rid of a lot of redundancies.